File: Work_leetcode/Contests/week 11/Convex Polygon.java
public class Solution {
    public boolean isConvex(List<List<Integer>> points) {
        if(points == null || points.size() <= 2) return true;
		points.add(points.get(0));
		points.add(points.get(1));
		
		boolean isNeg = false;
		for(int i = 2; i < points.size(); i++){
			int dx1 = points.get(i - 1).get(0) - points.get(i - 2).get(1);
			int dy1 = points.get(i - 1).get(1) - points.get(i - 2).get(1);
			int dx2 = points.get(i).get(0) - points.get(i - 1).get(0);
			int dy2 = points.get(i).get(1) - points.get(i - 1).get(1);
			
			long vector = dy2*dx1 - dy1*dx2;
			if(i == 2) isNeg = (vector < 0);
			else if(!(vector < 0 && isNeg || vector > 0 && !isNeg)) return false;
		}
		return true;
    }
}

File: Work_leetcode/Contests/week 11/Convex Polygon.cpp
class Solution {
public:
    bool isConvex(vector<vector<int>>& points) {
        int size = (int)points.size();
        assert(size >= 3);
        for (int i=0; i<size-1; ++i) {
            if (!judgeOneEdge(points, i, i+1))
                return false;
        }
        return true;
    }
    bool judgeOneEdge(vector<vector<int>>& points, int p1, int p2) {
        int size = (int)points.size();
        int x1 = points[p1][0], y1 = points[p1][1];
        int x2 = points[p2][0], y2 = points[p2][1];
        int a = y2-y1, b = x1 - x2, c = y1*(x2-x1) - x1*(y2 - y1);
        int x3, y3;
        if (p1 != 0 && p2 != 0) {
            x3 = points[0][0], y3 = points[0][1];
        } else if (p1 != 1 && p2 != 1) {
            x3 = points[1][0], y3 = points[1][1];
        } else {
            x3 = points[2][0], y3 = points[2][1];
        }
        int temp = a * x3 + b * y3 + c;
        if (temp < 0) {
            for (int i=0; i<size; ++i) {
                int x = points[i][0], y = points[i][1];
                if (a*x + b*y + c > 0) return false;
            }
        } else {
            for (int i=0; i<size; ++i) {
                int x = points[i][0], y = points[i][1];
                if (a*x + b*y + c < 0) return false;
            }
        }
        return true;
    }
};